utils/s2n_fork_detection.c

```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/* This captures Darwin specialities. This is the only APPLE flavor we care about.
+ * Here we also capture varius required feature test macros.
+ */
+#if defined(__APPLE__)
+    typedef struct _opaque_pthread_once_t  __darwin_pthread_once_t;
+    typedef __darwin_pthread_once_t pthread_once_t;
+    #define _DARWIN_C_SOURCE
+#elif !defined(_GNU_SOURCE)
+    /* Keep in sync with feature probe tests/features/madvise.c */
+    #define _GNU_SOURCE
```

Comment:
Can this "in sync" be enforced? Maybe related to implementing MADV_WIPEONFORK support detection via try_compile